broker/src/main/java/io/camunda/zeebe/broker/system/partitions/impl/steps/ZeebeDbPartitionTransitionStep.java

```diff
@@ -47,28 +51,48 @@
         || currentRole == Role.INACTIVE
         || context.getZeebeDb() == null) {
 
-      final ZeebeDb zeebeDb;
-      try {
-        context.getStateController().recover();
-        zeebeDb = context.getStateController().openDb();
-      } catch (final Exception e) {
-        Loggers.SYSTEM_LOGGER.error("Failed to recover from snapshot", e);
-
-        return CompletableActorFuture.completedExceptionally(
-            new IllegalStateException(
-                String.format(
-                    "Unexpected error occurred while recovering snapshot controller during leader partition install for partition %d",
-                    context.getPartitionId()),
-                e));
-      }
-
-      context.setZeebeDb(zeebeDb);
+      recoverDb(context, transitionFuture);
+    } else {
+      transitionFuture.complete(null);
     }
-    return CompletableActorFuture.completed(null);
+    return transitionFuture;
   }
 
   @Override
   public String getName() {
     return "ZeebeDb";
   }
+
+  private void recoverDb(
+      final PartitionTransitionContext context,
+      final CompletableActorFuture<Void> transitionFuture) {
+    final ActorFuture<Void> recoverFuture;
+    try {
+      recoverFuture = context.getStateController().recover();
+    } catch (final IOException error) {
+      transitionFuture.completeExceptionally(
+          String.format(RECOVERY_FAILED_ERROR_MSG, context.getPartitionId()), error);
+      return;
+    }
+
+    recoverFuture.onComplete(
+        (ok, error) -> {
+          if (error != null) {
+            transitionFuture.completeExceptionally(error);
+          } else {
+            try {
+              context.getStateController().verifyDb();
+              final var zeebeDb = context.getStateController().openDb();
+              context.setZeebeDb(zeebeDb);
+              transitionFuture.complete(null);
+            } catch (final Exception e) {
+              Loggers.SYSTEM_LOGGER.error("Failed to recover from snapshot", e);
+
+              transitionFuture.completeExceptionally(
+                  new IllegalStateException(
+                      String.format(RECOVERY_FAILED_ERROR_MSG, context.getPartitionId()), e));
+            }
```

Comment:
:x: why we have such duplicate code ?

Reply:
Which code is duplicated?

Reply:
Same as in broker/src/main/java/io/camunda/zeebe/broker/system/partitions/impl/steps/ZeebeDbPartitionStep.java

Reply:
Ah.. ok. `ZeebeDbPartitionStep` is the old transition code. I added the fix to that also in case we decided to turn the feature flag back due to the concurrent transitions bug. All classes named "*PartitionStep" is from the old transition and will be removed once we are confident with the new one.